data-contracts/README.md

```diff
@@ -0,0 +1,439 @@
+The decision is made to refactor the `membership` schema to encapsulate the date fields into a type - `ValidityPeriod` - which can be reused in other event types.
+
+```avroschema
+{
+    "type": "record",
+    "name": "ValidityPeriod",
+    "fields": [
+      {"name": "from", "type": {"type": "int", "logicalType": "date"}},
+      {"name": "to", "type": {"type": "int", "logicalType": "date"}
+      }
+    ]
+}
+```
+
+The refactored `membership` schema now references this new type:
+
+```avroschema
+{
+  "name": "Membership",
+  "namespace": "io.confluent.devrel",
+  "type": "record",
+  "fields": [
+    { "name": "user_id", "type": "string" },
+    { "name": "validity_period", "type": "io.confluent.devrel.ValidityPeriod" }
+  ]
+}
+```
+
+When we register version 2 of the `membership-avro-value` subject, use `addLocalReference` to include the `validityPeriod` type:
+
+```kotlin
+schemaRegistry{
+    register {
+        subject(inputSubject =  "membership-avro-value", type = "AVRO", file = "$avroSchemaDir/membership_v1.avsc")
+            .setMetadata("$metadataDir/membership_major_version_1.json")
+
+        subject(inputSubject =  "membership-avro-value", type = "AVRO", file = "$avroSchemaDir/membership_v2.avsc")
+            .addLocalReference("validity-period", "$avroSchemaDir/validity_period.avsc")
+            .setMetadata("$metadataDir/membership_major_version_2.json")
+            .setRuleSet("$rulesetDir/membership_migration_rules.json")
+    }
+}
+```
+
+Register the new schema using our gradle plugin:
+
+```shell
+./gradlew :data-contracts:schemas:registerSchemasTask
+```
+
+Download the schema to the version 2 application and generate the Java classes for this schema:
+
+```shell
+./gradlew :data-contracts:app-schema-v2:generateCode
+```
+
+In the `app-schema-v2` module, we'll find a new implementation of a `MembershipConsumer`:
+
+```kotlin
+class MembershipConsumer: BaseConsumer<String, Membership>(mapOf(
+    ConsumerConfig.GROUP_ID_CONFIG to "app-schema-v2",
+    ConsumerConfig.AUTO_OFFSET_RESET_CONFIG to "earliest",
+    ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG to "org.apache.kafka.common.serialization.StringDeserializer",
+    ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG to "io.confluent.kafka.serializers.KafkaAvroDeserializer",
+    AbstractKafkaSchemaSerDeConfig.LATEST_COMPATIBILITY_STRICT to true,
+    AbstractKafkaSchemaSerDeConfig.USE_LATEST_WITH_METADATA to "major_version=2"
+)
+) {
+    override fun consumeRecord(key: String, value: Membership) {
+        logger.info("v2 - Received Membership ${key}, ${value}")
+    }
+}
+```
+
+A closer look at the configuration of this consumer shows we are now using `latest.compatibility.strict` and specifying `major_version=2`
+in the `use.latest.with.metadata` configuration for the deserializer.
+
+Running the `main` function in `ApplicationV2Main` will consume the events from the `membership-avro` topic, but with a noticeable difference from the `app-schama-v1` application:
```

Comment:
similar to above - I wasn't actually able to run this due to the error I hit in IntelliJ